File: preprocess-uv-maps.js
/**
 * UV Map Preprocessor
 * Analyzes UV map images and creates pre-calculated mask data for each weapon part
 * Run this once: node preprocess-uv-maps.js
 */

const fs = require('fs');
const path = require('path');
const { createCanvas, loadImage } = require('canvas');

// Weapon parts we want to identify
const WEAPON_PARTS = ['barrel', 'magazine', 'body', 'stock', 'grip', 'sight', 'suppressor'];

// UV map filenames to process (matching weapon keys to UV files)
const WEAPON_UV_MAP = {
    'ak47': 'ak-47.png',
    'm4a4': 'm4a4.png',
    'm4a1s': 'm4a1-s.png',
    'awp': 'awp.png',
    'deagle': 'desert_eagle.png',
    'glock': 'glock-18.png',
    'usp': 'usp-s.png',
    'aug': 'aug.png',
    'bizon': 'bizon.png',
    'cz75': 'cz_75.png',
    'famas': 'famas.png',
    'fiveseven': 'five-seven.png',
    'g3sg1': 'g3sg1.png',
    'galil': 'galil_ar.png',
    'm249': 'm249.png',
    'mac10': 'mac-10.png',
    'mag7': 'mag-7.png',
    'mp5sd': 'mp5sd.png',
    'mp7': 'mp7.png',
    'mp9': 'mp9.png',
    'negev': 'negev.png',
    'nova': 'nova.png',
    'p2000': 'p2000.png',
    'p250': 'p250.png',
    'p90': 'p90.png',
    'revolver': 'revolver.png',
    'sawedoff': 'sawed-off.png',
    'scar20': 'scar-20.png',
    'sg553': 'sg_553.png',
    'ssg08': 'ssg_08.png',
    'tec9': 'tec-9.png',
    'ump45': 'ump-45.png',
    'xm1014': 'xm1014.png',
    'dualberettas': 'dual_berettas.png'
};

// Helper: Convert RGB to HSL
function rgbToHsl(r, g, b) {
    r /= 255;
    g /= 255;
    b /= 255;
    
    const max = Math.max(r, g, b);
    const min = Math.min(r, g, b);
    let h, s, l = (max + min) / 2;
    
    if (max === min) {
        h = s = 0;
    } else {
        const d = max - min;
        s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
        
        switch (max) {
            case r: h = ((g - b) / d + (g < b ? 6 : 0)) / 6; break;
            case g: h = ((b - r) / d + 2) / 6; break;
            case b: h = ((r - g) / d + 4) / 6; break;
        }
    }
    
    return [h * 360, s * 100, l * 100];
}

// Analyze which weapon part a pixel belongs to based on position and color
function identifyWeaponPart(x, y, width, height, r, g, b) {
    const normX = x / width;
    const normY = y / height;
    const [h, s, l] = rgbToHsl(r, g, b);
    
    // Skip completely black/transparent pixels
    if (l < 5) return null;
    
    // Part identification based on normalized position (UV coordinates)
    // These are optimized heuristics based on typical CS2 weapon UV layouts
    
    // Barrel - typically right side, horizontally aligned
    if (normX > 0.55 && normY > 0.35 && normY < 0.75) {
        return 'barrel';
    }
    
    // Stock - typically left side
    if (normX < 0.35 && normY > 0.35 && normY < 0.80) {
        return 'stock';
    }
    
    // Magazine - bottom center, often darker
    if (normX > 0.25 && normX < 0.60 && normY < 0.45 && l < 50) {
        return 'magazine';
    }
    
    // Sight - top center, often has metallic/bright areas
    if (normX > 0.40 && normX < 0.70 && normY > 0.70 && normY < 0.98) {
        return 'sight';
    }
    
    // Suppressor - far right, lighter tones
    if (normX > 0.80 && normY > 0.40 && normY < 0.70 && l > 40) {
        return 'suppressor';
    }
    
    // Grip - lower middle
    if (normX > 0.35 && normX < 0.55 && normY > 0.25 && normY < 0.50 && l > 20 && l < 65) {
        return 'grip';
    }
    
    // Body - center mass, everything else in middle region
    if (normX > 0.25 && normX < 0.75 && normY > 0.30 && normY < 0.85) {
        return 'body';
    }
    
    return 'body'; // Default to body if no specific match
}

// Compress mask data using Run-Length Encoding (RLE)
function compressMask(maskData, width, height) {
    const compressed = {};
    
    for (const part of WEAPON_PARTS) {
        const runs = [];
        let currentRun = null;
        
        for (let i = 0; i < maskData.length; i++) {
            const hasPart = maskData[i] === part;
            
            if (currentRun === null) {
                currentRun = { start: i, length: 1, value: hasPart };
            } else if (currentRun.value === hasPart) {
                currentRun.length++;
            } else {
                if (currentRun.value) {
                    runs.push([currentRun.start, currentRun.length]);
                }
                currentRun = { start: i, length: 1, value: hasPart };
            }
        }
        
        // Push last run
        if (currentRun && currentRun.value) {
            runs.push([currentRun.start, currentRun.length]);
        }
        
        if (runs.length > 0) {
            compressed[part] = runs;
        }
    }
    
    return compressed;
}

// Process a single UV map image
async function processUVMap(weaponKey, uvFileName) {
    console.log(`\n📋 Processing ${weaponKey} (${uvFileName})...`);
    
    const uvPath = path.join(__dirname, 'UVSheets', uvFileName);
    
    if (!fs.existsSync(uvPath)) {
        console.log(`  ⚠️  UV map not found: ${uvPath}`);
        return null;
    }
    
    try {
        const image = await loadImage(uvPath);
        const canvas = createCanvas(image.width, image.height);
        const ctx = canvas.getContext('2d');
        
        ctx.drawImage(image, 0, 0);
        const imageData = ctx.getImageData(0, 0, image.width, image.height);
        const pixels = imageData.data;
        
        console.log(`  📐 Image size: ${image.width}x${image.height}`);
        
        // Create mask array - store which part each pixel belongs to
        const maskData = new Array(image.width * image.height).fill(null);
        let pixelsCounted = 0;
        
        for (let y = 0; y < image.height; y++) {
            for (let x = 0; x < image.width; x++) {
                const i = (y * image.width + x) * 4;
                const r = pixels[i];
                const g = pixels[i + 1];
                const b = pixels[i + 2];
                const a = pixels[i + 3];
                
                // Skip fully transparent pixels
                if (a < 10) continue;
                
                const part = identifyWeaponPart(x, y, image.width, image.height, r, g, b);
                if (part) {
                    maskData[y * image.width + x] = part;
                    pixelsCounted++;
                }
            }
        }
        
        console.log(`  ✅ Analyzed ${pixelsCounted} pixels`);
        
        // Compress the mask data
        const compressed = compressMask(maskData, image.width, image.height);
        
        // Calculate statistics
        const stats = {};
        for (const part of WEAPON_PARTS) {
            const count = maskData.filter(p => p === part).length;
            if (count > 0) {
                stats[part] = count;
            }
        }
        
        console.log(`  📊 Part distribution:`, stats);
        
        return {
            weapon: weaponKey,
            width: image.width,
            height: image.height,
            masks: compressed,
            stats: stats,
            generated: new Date().toISOString()
        };
        
    } catch (error) {
        console.error(`  ❌ Error processing ${weaponKey}:`, error.message);
        return null;
    }
}

// Main processing function
async function main() {
    console.log('🚀 CS2 Skin Creator - UV Map Preprocessor');
    console.log('==========================================\n');
    
    // Create masks directory
    const masksDir = path.join(__dirname, 'masks');
    if (!fs.existsSync(masksDir)) {
        fs.mkdirSync(masksDir);
        console.log('📁 Created masks/ directory\n');
    }
    
    let processed = 0;
    let skipped = 0;
    
    // Process each weapon
    for (const [weaponKey, uvFileName] of Object.entries(WEAPON_UV_MAP)) {
        const result = await processUVMap(weaponKey, uvFileName);
        
        if (result) {
            // Save mask data as JSON
            const outputPath = path.join(masksDir, `${weaponKey}.json`);
            fs.writeFileSync(outputPath, JSON.stringify(result, null, 2));
            console.log(`  💾 Saved: masks/${weaponKey}.json`);
            processed++;
        } else {
            skipped++;
        }
    }
    
    console.log('\n==========================================');
    console.log(`✅ Preprocessing complete!`);
    console.log(`  Processed: ${processed} weapons`);
    console.log(`  Skipped: ${skipped} weapons`);
    console.log(`\n💡 Mask files saved to: ${masksDir}`);
    console.log(`\n🎯 Next step: Restart the app to use pre-calculated masks`);
}

main().catch(console.error);

File: preprocess-uv-maps.py
"""
UV Map Preprocessor for CS2 Skin Creator
Analyzes UV map images and creates pre-calculated mask data for each weapon part
Run: python preprocess-uv-maps.py
"""

import os
import json
from PIL import Image
from datetime import datetime
import colorsys

# Weapon parts to identify
WEAPON_PARTS = ['barrel', 'magazine', 'body', 'stock', 'grip', 'sight', 'suppressor']

# UV map filenames (matching weapon keys to UV files)
WEAPON_UV_MAP = {
    'ak47': 'ak-47.png',
    'm4a4': 'm4a4.png',
    'm4a1s': 'm4a1-s.png',
    'awp': 'awp.png',
    'deagle': 'desert_eagle.png',
    'glock': 'glock-18.png',
    'usp': 'usp-s.png',
    'aug': 'aug.png',
    'bizon': 'bizon.png',
    'cz75': 'cz_75.png',
    'famas': 'famas.png',
    'fiveseven': 'five-seven.png',
    'g3sg1': 'g3sg1.png',
    'galil': 'galil_ar.png',
    'm249': 'm249.png',
    'mac10': 'mac-10.png',
    'mag7': 'mag-7.png',
    'mp5sd': 'mp5sd.png',
    'mp7': 'mp7.png',
    'mp9': 'mp9.png',
    'negev': 'negev.png',
    'nova': 'nova.png',
    'p2000': 'p2000.png',
    'p250': 'p250.png',
    'p90': 'p90.png',
    'revolver': 'revolver.png',
    'sawedoff': 'sawed-off.png',
    'scar20': 'scar-20.png',
    'sg553': 'sg_553.png',
    'ssg08': 'ssg_08.png',
    'tec9': 'tec-9.png',
    'ump45': 'ump-45.png',
    'xm1014': 'xm1014.png',
    'dualberettas': 'dual_berettas.png'
}

def rgb_to_hsl(r, g, b):
    """Convert RGB to HSL color space"""
    r, g, b = r / 255.0, g / 255.0, b / 255.0
    h, l, s = colorsys.rgb_to_hls(r, g, b)
    return h * 360, s * 100, l * 100

def identify_weapon_part(x, y, width, height, r, g, b, a):
    """Identify which weapon part a pixel belongs to based on position and color"""
    
    # Skip transparent pixels
    if a < 10:
        return None
    
    norm_x = x / width
    norm_y = y / height
    h, s, l = rgb_to_hsl(r, g, b)
    
    # Skip very dark pixels (likely background)
    if l < 5:
        return None
    
    # Part identification based on normalized UV coordinates
    # These heuristics match typical CS2 weapon UV layouts
    
    # Barrel - typically right side, horizontally aligned
    if norm_x > 0.55 and 0.35 < norm_y < 0.75:
        return 'barrel'
    
    # Stock - typically left side
    if norm_x < 0.35 and 0.35 < norm_y < 0.80:
        return 'stock'
    
    # Magazine - bottom center, often darker
    if 0.25 < norm_x < 0.60 and norm_y < 0.45 and l < 50:
        return 'magazine'
    
    # Sight - top center, often metallic/bright
    if 0.40 < norm_x < 0.70 and 0.70 < norm_y < 0.98:
        return 'sight'
    
    # Suppressor - far right
    if norm_x > 0.80 and 0.40 < norm_y < 0.70 and l > 40:
        return 'suppressor'
    
    # Grip - lower middle
    if 0.35 < norm_x < 0.55 and 0.25 < norm_y < 0.50 and 20 < l < 65:
        return 'grip'
    
    # Body - center mass, default for middle region
    if 0.25 < norm_x < 0.75 and 0.30 < norm_y < 0.85:
        return 'body'
    
    return 'body'  # Default

def compress_mask_rle(mask_data, width, height):
    """Compress mask data using Run-Length Encoding"""
    compressed = {}
    
    for part in WEAPON_PARTS:
        runs = []
        current_run = None
        
        for i, pixel_part in enumerate(mask_data):
            has_part = (pixel_part == part)
            
            if current_run is None:
                current_run = {'start': i, 'length': 1, 'value': has_part}
            elif current_run['value'] == has_part:
                current_run['length'] += 1
            else:
                if current_run['value']:
                    runs.append([current_run['start'], current_run['length']])
                current_run = {'start': i, 'length': 1, 'value': has_part}
        
        # Add last run
        if current_run and current_run['value']:
            runs.append([current_run['start'], current_run['length']])
        
        if runs:
            compressed[part] = runs
    
    return compressed

def process_uv_map(weapon_key, uv_filename):
    """Process a single UV map image"""
    print(f"\n📋 Processing {weapon_key} ({uv_filename})...")
    
    uv_path = os.path.join('UVSheets', uv_filename)
    
    if not os.path.exists(uv_path):
        print(f"  ⚠️  UV map not found: {uv_path}")
        return None
    
    try:
        img = Image.open(uv_path).convert('RGBA')
        width, height = img.size
        pixels = img.load()
        
        print(f"  📐 Image size: {width}x{height}")
        
        # Create mask array
        mask_data = [None] * (width * height)
        pixels_counted = 0
        
        for y in range(height):
            for x in range(width):
                r, g, b, a = pixels[x, y]
                
                part = identify_weapon_part(x, y, width, height, r, g, b, a)
                if part:
                    mask_data[y * width + x] = part
                    pixels_counted += 1
        
        print(f"  ✅ Analyzed {pixels_counted:,} pixels")
        
        # Compress mask data
        compressed = compress_mask_rle(mask_data, width, height)
        
        # Calculate statistics
        stats = {}
        for part in WEAPON_PARTS:
            count = mask_data.count(part)
            if count > 0:
                stats[part] = count
        
        print(f"  📊 Part distribution: {stats}")
        
        return {
            'weapon': weapon_key,
            'width': width,
            'height': height,
            'masks': compressed,
            'stats': stats,
            'generated': datetime.now().isoformat()
        }
        
    except Exception as e:
        print(f"  ❌ Error processing {weapon_key}: {e}")
        return None

def main():
    print('🚀 CS2 Skin Creator - UV Map Preprocessor')
    print('=' * 50)
    print()
    
    # Create masks directory
    masks_dir = 'masks'
    if not os.path.exists(masks_dir):
        os.makedirs(masks_dir)
        print(f'📁 Created {masks_dir}/ directory\n')
    
    processed = 0
    skipped = 0
    
    # Process each weapon
    for weapon_key, uv_filename in WEAPON_UV_MAP.items():
        result = process_uv_map(weapon_key, uv_filename)
        
        if result:
            # Save mask data as JSON
            output_path = os.path.join(masks_dir, f'{weapon_key}.json')
            with open(output_path, 'w') as f:
                json.dump(result, f, indent=2)
            print(f"  💾 Saved: {output_path}")
            processed += 1
        else:
            skipped += 1
    
    print('\n' + '=' * 50)
    print(f'✅ Preprocessing complete!')
    print(f'  Processed: {processed} weapons')
    print(f'  Skipped: {skipped} weapons')
    print(f'\n💡 Mask files saved to: {masks_dir}/')
    print(f'\n🎯 Next: Restart app to use pre-calculated masks')

if __name__ == '__main__':
    main()
